Add reranking search results with MS Marco cross-encoder tutorial (#7634)
opensearch-project/documentation-website: _ml-commons-plugin/tutorials/reranking-cross-encoder.md

---
layout: default

> title: Reranking with MS Marco cross-encoder

Reviewer: ```suggestion
title: Reranking with the MS MARCO cross-encoder
```

parent: Tutorials
nav_order: 35
---

> # Reranking search results using the MS Marco cross-encoder model

Reviewer: ```suggestion
# Reranking search results using the MS MARCO cross-encoder model
```

A [reranking pipeline]({{site.url}}{{site.baseurl}}/search-plugins/search-relevance/reranking-search-results/) can rerank search results, providing a relevance score for each document in the search results with respect to the search query. The relevance score is calculated by a cross-encoder model. 

> This tutorial illustrates how to use the [Hugging Face Cross-Encoder for MS Marco](https://huggingface.co/cross-encoder/ms-marco-MiniLM-L-6-v2) model in a reranking pipeline. 

Author: ```suggestion
This tutorial illustrates how to use the [Hugging Face `ms-marco-MiniLM-L-6-v2` model](https://huggingface.co/cross-encoder/ms-marco-MiniLM-L-6-v2) in a reranking pipeline. 
```

Replace the placeholders beginning with the prefix `your_` with your own values.
{: .note}

## Prerequisite

Before you start, deploy the model on Amazon SageMaker. For better performance, use a GPU.

Run the following code to deploy the model on [Amazon SageMaker](https://aws.amazon.com/pm/sagemaker):

```python
import sagemaker
import boto3
from sagemaker.huggingface import HuggingFaceModel
sess = sagemaker.Session()
role = sagemaker.get_execution_role()

hub = {
    'HF_MODEL_ID':'cross-encoder/ms-marco-MiniLM-L-6-v2',
    'HF_TASK':'text-classification'
}
huggingface_model = HuggingFaceModel(
    transformers_version='4.37.0',
    pytorch_version='2.1.0',
    py_version='py310',
    env=hub,
    role=role, 
)
predictor = huggingface_model.deploy(
    initial_instance_count=1, # number of instances
    instance_type='ml.m5.xlarge' # ec2 instance type
)
```
{% include copy.html %}

Note the model inference endpoint; you'll use it to create a connector in the next step.

## Step 1: Create a connector and register the model

> First, create a connector for the model, provide the inference endpoint and your AWS credentials:

Reviewer 2: ```suggestion
First, create a connector for the model, providing the inference endpoint and your AWS credentials:
```

```json
POST /_plugins/_ml/models/_register?deploy=true
{
    "name": "Sagemaker Cross-Encoder model",
    "function_name": "remote",
    "description": "test rerank model",
    "connector_id": "your_connector_id"
}
```
{% include copy-curl.html %}

Note the model ID in the response; you'll use it in the following steps.

To test the model, call the Predict API:

```json
POST _plugins/_ml/models/your_model_id/_predict
{
  "parameters": {
    "inputs": [
      {
        "text": "I like you",
        "text_pair": "I hate you"
      },
      {
        "text": "I like you",
        "text_pair": "I love you"
      }
    ]
  }
}
```
{% include copy-curl.html %}

Each item in the `inputs` array comprises a `query_text` and a `text_docs` string, separated by a ` . `

Alternatively, you can test the model as follows:

```json
POST _plugins/_ml/_predict/text_similarity/your_model_id
{
  "query_text": "I like you",
  "text_docs": ["I hate you", "I love you"]
}
```
{% include copy-curl.html %}

The connector `pre_process_function` transforms the input into the format required by the `inputs` parameter shown in the previous Predict API request.

> By default, the SageMaker model output has the following format:

Reviewer: ```suggestion
By default, the SageMaker model output is in the following format:
```

```json
[
  {
    "label": "LABEL_0",
    "score": 0.054037678986787796
  },
  {
    "label": "LABEL_0",
    "score": 0.5877784490585327
  }
]
```

> The connector `pre_process_function` transforms the model output into the following format that the [rerank processor]({{site.url}}{{site.baseurl}}/search-plugins/search-pipelines/rerank-processor/) can interpret:

Reviewer: ```suggestion
The connector `pre_process_function` transforms the model output into the following format that can be interpreted by the [rerank processor]({{site.url}}{{site.baseurl}}/search-plugins/search-pipelines/rerank-processor/):
```

```json
{
  "inference_results": [
    {
      "output": [
        {
          "name": "similarity",
          "data_type": "FLOAT32",
          "shape": [
            1
          ],
          "data": [
            0.054037678986787796
          ]
        },
        {
          "name": "similarity",
          "data_type": "FLOAT32",
          "shape": [
            1
          ],
          "data": [
            0.5877784490585327
          ]
        }
      ],
      "status_code": 200
    }
  ]
}
```

> The response contains two `similarity` outputs. For each `similarity` output, the `data` array contains a relevance score of each document against the query. The `similarity` outputs are provided in the order of the input documents: the first similarity result pertains to the first document.

Reviewer: ```suggestion
The response contains two `similarity` outputs. For each `similarity` output, the `data` array contains a relevance score for each document against the query. The `similarity` outputs are provided in the order of the input documents: The first similarity result pertains to the first document.
```

## Step 2: Configure a reranking pipeline

Follow these steps to configure a reranking pipeline.

### Step 2.1: Ingest test data

Send a bulk request to ingest test data:

```json
POST _bulk
{ "index": { "_index": "my-test-data" } }
{ "passage_text" : "Carson City is the capital city of the American state of Nevada." }
{ "index": { "_index": "my-test-data" } }
{ "passage_text" : "The Commonwealth of the Northern Mariana Islands is a group of islands in the Pacific Ocean. Its capital is Saipan." }
{ "index": { "_index": "my-test-data" } }
{ "passage_text" : "Washington, D.C. (also known as simply Washington or D.C., and officially as the District of Columbia) is the capital of the United States. It is a federal district." }
{ "index": { "_index": "my-test-data" } }
{ "passage_text" : "Capital punishment (the death penalty) has existed in the United States since beforethe United States was a country. As of 2017, capital punishment is legal in 30 of the 50 states." }
```
{% include copy-curl.html %}

### Step 2.2: Create a reranking pipeline

> Create a reranking pipeline with the MS Marco cross-encoder model:

Reviewer: ```suggestion
Create a reranking pipeline using the MS MARCO cross-encoder model:
```

```json
PUT /_search/pipeline/rerank_pipeline_sagemaker
{
    "description": "Pipeline for reranking with Sagemaker cross-encoder model",
    "response_processors": [
        {
            "rerank": {
                "ml_opensearch": {
                    "model_id": "your_model_id_created_in_step1"
                },
                "context": {
                    "document_fields": ["passage_text"]
                }
            }
        }
    ]
}
```
{% include copy-curl.html %}

> If you provide multiple field names in `document_fields`, the values of all fields are first concatenated and then reranking is performed.

Reviewer: ```suggestion
If you provide multiple field names in `document_fields`, then the values of all fields are first concatenated, after which reranking is performed.
```
